docs/src/cookbooks/hierarchy.md

```diff
@@ -0,0 +1,171 @@
+
+`Definition(...)` enables elaborating a module, but does not actually instantiate that module. Instead, it returns a `Definition` class which represents that module's definition.
+
+`Instance(...)` takes a `Definition` and instantiates it, returning an `Instance` object.
+
+Modules (classes or traits) which will be used with the `Definition`/`Instance` api should be marked with the `@instantiable` annotation at the class/trait definition.
+
+To make a Module's members variables accessible from an `Instance` object, they must be annotated with the `@public` annotation. Note that this is only accessible from a Scala sense -- this is not in and of itself a mechanism for cross-module references.
+
+In the following example, use `Definition`, `Instance`, `@instantiable` and `@public` to create multiple instances of one specific parameterization of a module, `AddOne`.
+
+```scala mdoc:reset
+import chisel3._
+import chisel3.experimental.hierarchy._
+
+@instantiable
+class AddOne(width: Int) extends Module {
+  @public val in  = IO(Input(UInt(width.W)))
+  @public val out = IO(Output(UInt(width.W)))
+  out := in + 1.U
+}
+
+class AddTwo(width: Int) extends Module {
+  val in  = IO(Input(UInt(width.W)))
+  val out = IO(Output(UInt(width.W)))
+  val addOneDef = Definition(new AddOne(width))
+  val i0 = Instance(addOneDef)
+  val i1 = Instance(addOneDef)
+  i0.in := in
+  i1.in := i0.out
+  out   := i1.out
+}
+
+chisel3.stage.ChiselStage.emitVerilog(new AddTwo(10))
+```
+
+## How do I access internal fields of an instance?
+
+You can mark internal members of a class or trait marked with `@instantiable` with the `@public` annotation. The requirements are that the field is publically accessable, is a val or lazy val, and is a valid type.
+The list of valid types are:
+
+1. IsInstantiable
+2. IsLookupable
+3. Data
+4. BaseModule
+5. Iterable/Option containing a type that meets these requirements
+6. Basic type like String, Int, BigInt etc.
+
+To mark a superclass's member as `@public`, use the following pattern (shown with `val clock`).
+
+```scala mdoc:silent:reset
+import chisel3._
+import chisel3.experimental.hierarchy._
+
+@instantiable
+class MyModule extends Module {
+  @public val clock = clock
+}
+```
+
+You'll get the following error message for improperly marking something as `@public`:
+
+```scala mdoc:reset:fail
+import chisel3._
+import chisel3.experimental.hierarchy._
+
+object NotValidType
+
+@instantiable
+class MyModule extends Module {
+  @public val x = NotValidType
+}
+```
+
+## How do I make my parameters accessable from an instance?
+
+If an instance's parameters are simple (e.g. Int, String etc.) they can be marked directly with `@public`.
+
+Oftentimes parameters are more complicated and are contained in case classes. In these cases, mark the case class with the `IsLookupable` trait. This means that this type can be returned directly from an instance.
```

Comment:
What does "returned directly from an instance"? Does this mean "accessed"? I'm being pedantic here as "return" implies something involving a function and results.